### ISSUE TITLE
Calculate max vDisk size based on storage deployment configuration

### ISSUE THREAD

**Issue by FastGeert** [0-complexity/openvcloud#1122]
Needs to be applied to vdisk creation and resize.

See https://github.com/0-complexity/openvcloud/issues/1119#issuecomment-352678785 and below

**Comment by grimpy** [comment]
I would do the following:
if block size is X then max size is X

    4k block size= max 2TB
    8K block size= max 4TB
    16K block size= max 8TB
    32K block size= max 16TB
    64K block size= max 32TB
    128K block size= max 64TB

To retrieve the block size = look at a get of the Vpool/id
@grimpy grimpy added the state_verification label 4 hours ago
@wvandebroeck
Member
wvandebroeck commented 2 hours ago

Lets make it a little better than the above
get the block size of the VPool (cluster_size': is the variable you need, in this case its 4):

n [5]: client.get('/vpools/7b60192f-11c1-438b-a8d8-3ae193818223')
Out[5]:
{u'configuration': {u'cluster_size': 4,
u'dtl_config_mode': u'Automatic',
u'dtl_enabled': True,
u'dtl_mode': u'sync',
u'dtl_transport': u'tcp',
u'mds_config': {u'mds_maxload': 75, u'mds_safety': 3, u'mds_tlogs': 100},
u'sco_size': 4,

How to determine the max VDISK size from this paramater:
Formula to calculate the max Vdisk size 2TB * cluster_size/4
in this case: 2 * 4/4 = 2TB is the max size of a Vdisk in this pool.

**Comment by grimpy** [comment]
this isnt 100% correct we also need to considuer the cache_ratio, but the ides is possible yes.

**Comment by wvandebroeck** [comment]
@grimpy great lets include all things we need to get this solid.

**Comment by wimpers** [comment]
Hi @grimpy 

I don't see why a bigger disk would have any influence in the (metadata ) cache ratio. A vDisks currently has 8 bytes for addressing its LBAs irrespective of the blocksize. These 8 bytes get stored in the page cache. 

Of course there will be an impact on the SSD cache backend.  If you set the cache ratio on the SSD backend to 20GB and your blocksize is 4K (so max 2TB disk), the percentage of the disk you cache is much bigger than if you would use a higher blocker size.

Met me know if this would not be clear.

**Comment by grimpy** [comment]
@wimpers As far as I can recall from our sessions it's not about a hard limit but more about performance and recommended values. If we increase the disk to a bigger size with a certain cache_ration the memory in the volumedriver would exceed the recommended values which would have as sideeffect that we could serve less vdisks from this volumedriver. 

Let me know if I got this wrong.

So as far as I remember the recommended values was 2TB with 4K block with 100% cache_ratio
So the formula would be: (512MiB * Blocksize) / (cache_ratio)

**Comment by wimpers** [comment]
> If we increase the disk to a bigger size with a certain cache_ration the memory in the volumedriver 

This is correct. One side remark the amount of data in mem wil be the same for 
* 2TB disk, 4K blocksize, 50% md ratio
* 4TB disk, 8K blocksize, 50% md ratio 

But do note that the question was around max size of a vDisk and how this related to the blocksize.

> So as far as I remember the recommended values was 2TB with 4K block with 100% cache_ratio

For best performance it is recommended to keep 100% in mem. This has an impact on memory cost so you might want to set it to  something lower.

> So the formula would be: (512MiB * Blocksize) / (cache_ratio)

I'm don't understand this one. What are you trying to calculate? The amount of MD in mem? If so it would be cache_ratio * (vidsk size / blocksize) * 8 byte (see calculator)

**Comment by grimpy** [comment]
@wimpers the formula I mentioned is for calculating the max vDisk size that works with our hardware.
Since we originally designed the memory for a fixed amount of vDisks and memory.

**Comment by wimpers** [comment]
@grimpy 

sorry it is Monday morning ... I still don't understand the formula. Where does the 512MiB come from?

**Comment by grimpy** [comment]
@wimpers good question, I think we have been told that with our amount of memory per storage node we can serve upto X amount of vdisks of size 2TiB with 4KiB block and 100% cacheratio. I dont know how much memory we have and I dont know what value is X either.

So I just used the same ratio to create the formula
2TiB/4KiB == 512MiB

**Comment by wimpers** [comment]
@grimpy ok now I get it. Do note that TiB/KiB doesn't result in MiB but in a  number.
